cloud/previews/previews.go

```diff
@@ -0,0 +1,178 @@
+	Stack *config.Stack
+	Cmd   []string
+}
+
+// CreatePreviewOpts is the options for the CreatePreview function
+type CreatePreviewOpts struct {
+	Runs            []RunContext
+	AffectedStacks  map[string]*config.Stack
+	OrgUUID         cloud.UUID
+	UpdatedAt       int64
+	Technology      string
+	TechnologyLayer string
+	Repository      string
+	DefaultBranch   string
+	ReviewRequest   *cloud.DeploymentReviewRequest
+	Metadata        *cloud.DeploymentMetadata
+}
+
+// CreatedPreview is the result of the CreatePreview function
+type CreatedPreview struct {
+	ID                    string
+	StackPreviewsByMetaID map[string]string
+}
+
+// ChangesetDetails is the details of a changeset
+type ChangesetDetails struct {
+	Provisioner    string
+	ChangesetASCII string
+	ChangesetJSON  string
+}
+
+// CreateStackPreviewOpts is the options for the CreateStackPreview function
+type CreateStackPreviewOpts struct {
+	OrgUUID          cloud.UUID
+	StackPreviewID   string
+	ChangesetDetails *ChangesetDetails
+}
+
+// StackPreviewLogLine is a log line for a stack preview
+type StackPreviewLogLine struct {
+	LogLine   string
+	Timestamp time.Time
+	Channel   string
+	Message   string
+}
+
+// AppendStackPreviewLogsOpts is the options for the AppendStackPreviewLogs function
+type AppendStackPreviewLogsOpts struct {
+	OrgUUID        cloud.UUID
+	StackPreviewID string
+	Logs           []StackPreviewLogLine
+}
+
+// AppendStackPreviewLogs appends logs to a stack preview in the cloud
+// nolint:revive
+func AppendStackPreviewLogs(client *cloud.Client, timeout time.Duration, opts AppendStackPreviewLogsOpts) (map[string]string, error) {
+	return nil, errors.E("not implemented")
+}
+
+// CreateStackPreview creates a new stack preview in the cloud
+// nolint:revive
+func CreateStackPreview(client *cloud.Client, timeout time.Duration, opts CreateStackPreviewOpts) (map[string]string, error) {
+	return nil, errors.E("not implemented")
+}
+
+// CreatePreview creates a new preview in the cloud
+func CreatePreview(client *cloud.Client, timeout time.Duration, opts CreatePreviewOpts) (*CreatedPreview, error) {
+	if err := opts.validate(); err != nil {
+		return nil, err
+	}
+
+	payload := cloud.CreatePreviewPayloadRequest{
+		UpdatedAt:       opts.UpdatedAt,
+		Technology:      opts.Technology,
+		TechnologyLayer: opts.TechnologyLayer,
+		ReviewRequest:   opts.ReviewRequest,
+		Metadata:        opts.Metadata,
+		Stacks:          []cloud.PreviewStack{},
+	}
+
+	previewStacksMap := map[string]RunContext{}
+	for _, run := range opts.Runs {
+		previewStacksMap[run.Stack.ID] = run
+	}
+
+	// loop over all affected stacks, if an item is present in the
+	// previewStacksMap, use the preview status and cmd from there
+	for _, affectedStack := range opts.AffectedStacks {
+		tags := []string{}
+		if len(affectedStack.Tags) > 0 {
+			tags = affectedStack.Tags
+		}
+
+		stack := cloud.PreviewStack{
+			PreviewStatus:   PreviewStatusAffected,
+			Repository:      opts.Repository,
```

Comment:
this is not validated. Is repo okay to be empty?

Reply:
its validated here: https://github.com/terramate-io/terramate/blob/22329f6c6657eb9afe9462f48236bb1053eea6b8/cloud/types_previews.go#L75

but I will change that to use the existing `Stacks` validation to remove the duplication.